+str #17579 #17617 EventAdapter

ktoso opened the pull request:
Initial PR for <strike>ModelAdapter</strike> EventAdapter.

Noteworthy:
- optionally takes ActorSystem as the adapter may need the config or serializers etc
- one adapter per journal
- instead of identity adapter currently avoiding to call it if not set, did not benchmark if it matters or not

refs #17579 #17617

ktoso committed:
+per #17579 #17617 Introduces EventAdapter
+ per plugin scoped adapters
+ could be swapped during runtime
+per EventAdapter now has manifest and is configurable ai la serializers
+ json examples in docs
+ including "completely manual" example in case one wants to add
  metadata TO the persisted event
+ better error reporting when misconfigured bindings
+ manifest is handled by in memory plugin
- did not check if it works with LevelDB plugin yet
> TODO: json example uses Gson, as that's simplest to do, can we use
+per allows 1:n adapters, multiple adapters can be bound to 1 class

patriknw commented on akka-persistence/src/main/scala/akka/persistence/JournalProtocol.scala:
```diff
@@ -46,6 +48,13 @@ private[persistence] object JournalProtocol {
+   * Register an model adapter for the Journal
+   * @param adapter to be applied on all events going to this journal
+   */
+  final case class RegisterModelAdapter(adapter: ModelAdapter[Any])
```
naming: EventAdapter?

ktoso replied:
+1, sounds good

patriknw commented on akka-persistence/src/main/scala/akka/persistence/journal/AsyncWriteJournal.scala:
```diff
@@ -47,7 +47,9 @@ trait AsyncWriteJournal extends Actor with WriteJournalBase with AsyncRecovery {
       asyncReplayMessages(persistenceId, fromSequenceNr, toSequenceNr, max) { p ⇒
-        if (!p.deleted || replayDeleted) persistentActor.tell(ReplayedMessage(p), p.sender)
+        if (!p.deleted || replayDeleted) {
+          persistentActor.tell(ReplayedMessage(adaptFromJournal(p)), p.sender)
```
I think we also need to provide a type hint to `adaptFromJournal`, similar the string `manifest` for serializers

ktoso replied:
Right, needed for "I do my mapping manually"

patriknw commented on akka-persistence/src/main/scala/akka/persistence/Persistence.scala:
```diff
@@ -215,6 +253,34 @@ class Persistence(val system: ExtendedActorSystem) extends Extension {
+    val pluginConfig = system.settings.config.getConfig(configPath)
+    if (!pluginConfig.hasPath("adapter")) None
+    else {
+      val adapterClassName = pluginConfig.getString("adapter")
```
I might have completely misunderstood your "tagging prototype", but I thought you had something similar to the serialization extension that would support one tagger/adapter per configured event type? I thought that would be nice, but this is perhaps enough?

ktoso replied:
Oh - no I think you got me right, see here for tagger impl: https://github.com/ktoso/akka/commit/410a4a8738296187076d1b74074a6ee15798c8a4#diff-aa1e901f1dd62bc57d4b969106f64c27R70 It's as serializers "pick most specific one" and allowed for multiple ones to be combined into one (that part we do not need for adapters).

I somehow thought "one adapter per journal", but now that you mention it "serializer like registration" for those makes much sense, so I'll bring back that idea. :)

patriknw replied:
yes, and that's why I wanted to combine the adapter with the tagger, so that user doesn't have to configure (mostly) same thing twice

drewhk commented:
the comment tag "str" is wrong :)

ktoso commented on akka-persistence/src/test/scala/akka/persistence/EventAdapterSpec.scala:
```diff
@@ -0,0 +1,204 @@
+         |    "${classOf[EventAdapterSpec].getCanonicalName}$$TaggedDataChanged" = age
+         |    "${classOf[EventAdapterSpec].getCanonicalName}$$MyTagJournal$$JournalModel"      = unwrap
+         |  }
+         |
```
@patriknw so that's the style of configuration we can do:

- adapters defined in one place
- bindings defined at top level, or:
- for each plugin separately

ktoso replied:
(this test fails now, I messed up something with when which adapter is used)

patriknw replied:
I think it is easiest to understand with one config section for each plugin, i.e. 

    akka.persistence.journal.leveldb {
      event-adapters { ... }
      event-adapter-bindings { ... }
    }

patriknw commented on akka-docs/rst/scala/code/docs/persistence/PersistenceEventAdapterDocSpec.scala:
```diff
@@ -0,0 +1,195 @@
+          }
+
+          event-adapter-bindings {
+            "docs.persistence.MySnapshot" = my-snapshot
```
where is `my-snapshot` defined?

ktoso replied:
bug, thanks!

patriknw commented on akka-docs/rst/scala/code/docs/persistence/PersistenceEventAdapterDocSpec.scala:
```diff
@@ -0,0 +1,195 @@
+  val gson = new Gson // TODO: serisouly, simplest to plug in, all type-class based ones fail in this scenario
+
+  override val includeManifest = false // we take care of it manually
+  override def manifest(event: Any): String = event.getClass.getCanonicalName
```
why not always include the manifest, i.e. remove the `def includeManifest`? if not needed it can be `""`

ktoso replied:
Sounds good

patriknw commented on akka-persistence/src/main/scala/akka/persistence/Persistent.scala:
```diff
@@ -63,6 +64,11 @@ trait PersistentRepr extends PersistentEnvelope with Message {
   /**
+   * Creates a new persistent message with the specified `manifest`.
+   */
+  def withManifest(payload: String): PersistentRepr
```
s/payload/manifest/

ktoso replied:
thanks!

patriknw commented on akka-persistence/src/main/scala/akka/persistence/journal/EventAdapter.scala:
```diff
@@ -0,0 +1,57 @@
+   * Return the manifest (type hint) that will be provided in the `fromJournal` method.
+   * Use `""` if manifest is not needed.
+   */
+  def manifest(event: Any): String = ""
```
I think this should be an abstract method, i.e. no default impl

ktoso replied:
ok :+1: and removed the `includeManifest method.

patriknw commented on akka-persistence/src/main/scala/akka/persistence/journal/WriteJournalBase.scala:
```diff
@@ -4,21 +4,36 @@
+
+  /** INTERNAL API */
+  private[akka] final def adaptFromJournal(repr: PersistentRepr): PersistentRepr = {
+    val adapter = eventAdapters.get(repr.payload.getClass)
```
Something is missing here. The stored format is typically the same for all events, e.g. String, Json, but there is no way to map that class to different adapters. I think something similar to the serializerId is needed, i.e. the id is stored and on the way back the adapter is looked up by id instead of by class.

ktoso replied:
Right, I see what you mean - working on it.

ktoso replied:
See https://github.com/ktoso/akka/commit/29d82b1875ab9a6639bb96bc7a25015f3d8c2714 for the "adapterId" implementation. Today we figured out during the meeting that even this is not enough, as it does not cover:

- initially no adapter
- now we need to add an adapter for some type
- that type could be "json", so it's hard to restrict to one

patriknw replied:
What is the exact use case for that? Isn't that overkill? How did you create json if you didn't use an adapter in the first place?

ktoso replied:
Enjoy the midsommar! :wink: (over here it's raining) (no releases today, no worries)

One of the possible scenarios is (it also starts to play into the ideas needed for https://github.com/akka/akka/issues/17616):

"old app"
- `persist(OldAppEvent)`
- catch all "toJson" adapter
  - stored id is something, lets say 1001

We start the new version of the app:
- "we need to move away from OldAppEvents!"
  - they are all stored as json, that's what the adapter gets
  - the model has become complex and we want multiple adapters for different types that are "inside the json"
  - the adapter gets json to the fromJournal() - current impl it would for all of these have to be the 1001 adapter
- **new from here-on:** we add EventAdapters with `fromJournal: Json => NewAppEvent`
  - adapter knows based on something inside the json if it should emit events or not (we provide `EventSeq.empty`)
  - can emit "nothing"
  - allow 1:n which we did not yet address

How it could be done with current:
- was stored without adapter so adapterId is 0
- users are kind of stuck with that decision if we only allow picking the adapter by the adapter id, we could allow binding to the 0 adapterId (which in theory should be our "internal reserved id" for identity)...

patriknw commented on akka-persistence/src/test/scala/akka/persistence/EventAdapterSpec.scala:
```diff
@@ -0,0 +1,217 @@
+         |    age                 = "${classOf[EventAdapterSpec].getCanonicalName}$$UserAgeTaggingAdapter"
+         |    replay-pass-through = "${classOf[EventAdapterSpec].getCanonicalName}$$ReplayPassThroughAdapter"
+         |    unwrap              = "${classOf[EventAdapterSpec].getCanonicalName}$$PayloadUnwrapper"
+         |  }
```
I'm not sure we should complicate it with these two possible locations of `event-adapters`.
It is possible to extract common definitions with typesafe config anyway using syntax like:

    inmem.event-adapters = ${common-event-adapters}

ktoso replied:
ok and concating lists works as well... simplifying now

patriknw commented on project/Dependencies.scala:
```diff
@@ -36,6 +36,11 @@ object Dependencies {
 
+    object Docs {
+      val sprayJson   = "io.spray"                   %%  "spray-json"                  % "1.3.2"
+      val gson        = "com.google.code.gson"        % "gson"                         % "2.3.1"
```
"test" scope?

ktoso replied:
thanks!

patriknw commented:
This is in the right direction. I don't think we can get away from adding an `adapterId` also.

patriknw commented on akka-persistence/src/main/scala/akka/persistence/journal/AsyncWriteProxy.scala:
```diff
@@ -29,10 +29,13 @@ private[persistence] trait AsyncWriteProxy extends AsyncWriteJournal with Stash
 
   override def receive = {
     case SetStore(ref) ⇒
+      println("SetStore(ref) = " + SetStore(ref) + ", " + context.system)
```
change to log.debug?

patriknw commented on akka-persistence/src/main/scala/akka/persistence/journal/AsyncWriteProxy.scala:
```diff
@@ -29,10 +29,13 @@ private[persistence] trait AsyncWriteProxy extends AsyncWriteJournal with Stash
       context.become(initialized)
-    case _ ⇒ stash()
+    case x ⇒
+      println("x = " + x)
```
remove or debug?

ktoso replied:
Oh, actually will remove here - somewhere on this branch I broke `SharedLeveldbJournalSpec`, this is from investigating it

patriknw commented on akka-persistence/src/main/scala/akka/persistence/journal/SyncWriteJournal.scala:
```diff
@@ -41,11 +41,16 @@ trait SyncWriteJournal extends Actor with WriteJournalBase with AsyncRecovery {
       } recover {
-        case e ⇒ ReplayMessagesFailure(e)
+        case e ⇒
+          e.printStackTrace()
```
remove

patriknw commented:
@ktoso noticed that you updated, is it ready for review?

ktoso commented:
Yes I think you can give it a read, I know it breaks `SharedLeveldbJournalSpec` however have not figured out why it does yet (so looking into it now).

ktoso commented on akka-docs/rst/scala/code/docs/persistence/PersistenceEventAdapterDocSpec.scala:
```diff
@@ -0,0 +1,233 @@
+    }
+  }
+}
+
```
Docs here are incomplete (impls not included in docs, but that will be part of the documenting schema evolution ticket)

patriknw commented on akka-docs/rst/java/persistence.rst:
```diff
@@ -503,6 +503,53 @@ not accept more messages and it will throw ``AtLeastOnceDelivery.MaxUnconfirmedM
+
+Event Adapters help in situations where:
+
+- **Upcasting** – existing events stored in *Version 1* should be "upcasted" to a new *Version 2* representation,
```
I really dislike the upcasting term. I know where it comes from, but it doesn't feel right (and requires explanation). I prefer migration.

ktoso replied:
Ok, how about we call it "model migrations" here? I'd like to add the word upcasting in something like "This technique is sometimes refered to as 'upcasting' in some other libraries"? In case someone familiar with Axon or something like it tries to find "how do I do upcasting in akka?"

patriknw replied:
"version migrations"?
(mentioning upcasting is fine)

patriknw commented on akka-docs/rst/java/persistence.rst:
```diff
@@ -503,6 +503,53 @@ not accept more messages and it will throw ``AtLeastOnceDelivery.MaxUnconfirmedM
+  the ``toJournal`` function is usually an identity function, however the ``fromJournal`` is implemented as
+  ``v1.Event=>v2.Event``, performing the neccessary upcasting inside the fromJournal method.
+- **Separating Domain and Data models** – thanks to EventAdapters it is possible to completely separate the domain model
+  from the model used to persist data in the Journals. For example one may want to use nice to use case classes in the
```
had to read this several times "to use nice to use".
I think this is enough: For example one may want to use case classes...

ktoso replied:
agreed, thanks :)

patriknw commented on akka-docs/rst/java/persistence.rst:
```diff
@@ -503,6 +503,53 @@ not accept more messages and it will throw ``AtLeastOnceDelivery.MaxUnconfirmedM
+  from the model used to persist data in the Journals. For example one may want to use nice to use case classes in the
+  domain model, however persist their protocol-buffer (or any other binary serialization format) counter-parts to the Journal.
+  A simple ``toJournal:MyModel=>MyDataModel`` and ``fromJournal:MyDataModel=>MyModel`` adapter can be used to implement this feature.
+- **Journal Specialized Data Types** – exposing data types understood by the underlying Journal, for example for data stores which understand JSON its possible
```
it's

patriknw commented on akka-docs/rst/java/persistence.rst:
```diff
@@ -503,6 +503,53 @@ not accept more messages and it will throw ``AtLeastOnceDelivery.MaxUnconfirmedM
+
+.. includecode:: ../scala/code/docs/persistence/PersistenceEventAdapterDocSpec.scala#event-adapters-config
+
+It is possible to bind multiple adapters to one class, in which case the event will be given (in order of definition in the configuration)
```
perhaps clarify that this is only for `fromJournal`

ktoso replied:
:+1:

patriknw commented on akka-persistence/src/main/scala/akka/persistence/Persistent.scala:
```diff
@@ -95,9 +101,10 @@ object PersistentRepr {
     deleted: Boolean = false,
     sender: ActorRef = null): PersistentRepr =
-    PersistentImpl(payload, sequenceNr, persistenceId, deleted, sender)
+    PersistentImpl(payload, sequenceNr, persistenceId, manifest, /*adapterId,*/ deleted, sender)
```
remove

ktoso replied:
thanks!

ktoso commented on akka-persistence/src/test/scala/akka/persistence/EndToEndEventAdapterSpec.scala:
```diff
@@ -0,0 +1,241 @@
+
+  "EventAdapters in end-to-end scenarios" must {
+
+    "use the same adapter when reading as was used when writing to the journal, by its adapterId" in
```
fixed test title not to include adapterId now, push soon

patriknw commented on akka-persistence/src/main/scala/akka/persistence/journal/EventAdapter.scala:
```diff
@@ -0,0 +1,91 @@
+   * may require adapters to return `com.example.myjournal.EventWithPrimaryKey(event, key)`.
+   *
+   * The `toJournal` adaptation must be an 1-to-1 transformation.
+   * It is now allowed to drop incoming events during the `toJournal` adaptation.
```
s/now/not/ ?

ktoso replied:
:+1: thanks

patriknw commented on akka-persistence/src/main/scala/akka/persistence/journal/EventAdapter.scala:
```diff
@@ -0,0 +1,91 @@
+  final def apply(events: Any*): EventSeq = EventsSeq(events.toList)
+}
+final case class SingleEventSeq(event: Any) extends EventSeq { // TODO try to make it a value class, would save allocations
+  override def events: immutable.Seq[Any] = List(event)
```
val

ktoso replied:
Thanks. 
I was trying to make this one a value class, however things get complicated then sadly...
(`EventSeq` must be an interface and must `extend Any` (universal trait), which destroys the Java API forwarders etc). I hope to try this again, we could save on some allocations here.

patriknw commented on akka-persistence/src/main/scala/akka/persistence/journal/WriteJournalBase.scala:
```diff
@@ -4,16 +4,39 @@
-      case p: PersistentRepr ⇒ p.update(sender = Actor.noSender) // don't store sender
+    rb.collect { // collect instead of flatMap to avoid Some allocations
+      case p: PersistentRepr ⇒
+        p.update(sender = Actor.noSender) // don't store sender
```
it is immutable, so you must pass in this result in next stmt

    adaptToJournal(p.update(sender = Actor.noSender))

ktoso replied:
whoops, silly rebase mistake; thanks!

patriknw commented:
LGTM, aside from some minor things

ktoso commented:
Addressed comments and fixed bug in LevelDBStore which failed validation previously.
I'll merge once Validation passes.

More work around docs for this coming soon anyway, but in next Milestone.

patriknw commented on akka-persistence/src/main/scala/akka/persistence/journal/leveldb/LeveldbStore.scala:
```diff
@@ -112,11 +111,11 @@ class SharedLeveldbStore extends { val configPath = "akka.persistence.journal.le
 
   def receive = {
-    case WriteMessages(msgs)                        ⇒ sender() ! writeMessages(msgs)
+    case WriteMessages(msgs)                        ⇒ sender() ! writeMessages(preparePersistentBatch(msgs))
```
great catch

ktoso committed:
+pro #17611 add perl script to find javadoc errors

ktoso commented:
Fails here on JavaDoc which I wasn't able to figure out precisely yet.

ktoso commented:
Pushed the javadoc fixup, should be OK to merge once validation confirms its happy.

patriknw commented:
thanks @ktoso, I hope you did that from bed ;-)

ktoso commented:
Correct, after a round of mid-day sleeping ;-) Happy to catch up a bit (no pressure) here and there.
